File: scip-java/src/main/scala/com/sourcegraph/scip_java/buildtools/ClasspathEntry.scala
package com.sourcegraph.scip_java.buildtools

import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

import scala.annotation.tailrec
import scala.jdk.CollectionConverters._

import com.sourcegraph.scip_semanticdb.MavenPackage

/**
 * Represents a single classpath entry on the classpath of a project, used to
 * emit SCIP "packageInformation" nodes. A classpath entry can either be a jar
 * file or a directory path.
 */
case class ClasspathEntry(
    entry: Path,
    sources: Option[Path],
    groupId: String,
    artifactId: String,
    version: String
) {
  def toPackageHubId: String = s"maven:$groupId:$artifactId:$version"
  def toPackageInformation: MavenPackage =
    new MavenPackage(entry, groupId, artifactId, version)
}

object ClasspathEntry {

  /**
   * Parses ClasspathEntry from the SemanticDB targetroot directory.
   *
   * Two separate formats are supported:
   *
   *   - javacopts.txt: line-separated list of Java compiler options.
   *   - dependencies.txt: line-separated list of dependency information.
   *
   * Note that the targetroot can contain several files with names ending in
   * "dependencies.txt" - for example if they come from a multi-module build.
   *
   * @param targetroot
   *   @return
   */
  def fromTargetroot(
      targetroot: Path,
      sourceroot: Path
  ): List[ClasspathEntry] = {
    val javacopts = targetroot.resolve("javacopts.txt")
    if (Files.isRegularFile(javacopts))
      fromJavacopts(javacopts, sourceroot)
    else
      discoverDependenciesFromFiles(targetroot)
  }

  /**
   * Discover all files that end in "dependencies.txt" directly under
   * targetroot. There can be many files because we will be writing dependencies
   * for multiple projects.
   *
   * @param targetroot
   *   @return classpath entries read from the discovered files
   */
  private def discoverDependenciesFromFiles(
      targetroot: Path
  ): List[ClasspathEntry] = {
    os.list
      .stream(os.Path(targetroot))
      .filter(p => os.isFile(p) && p.last.endsWith("dependencies.txt"))
      .map(path => fromDependencies(path.toNIO))
      .toList
      .flatten
      .distinct
  }

  /**
   * Parses ClasspathEntry from a "dependencies.txt" file
   *
   * Every line of the file is a tab separated value with the following columns:
   * groupId, artifactId, version, path to the jar file OR classes directory
   * path.
   */
  private def fromDependencies(dependencies: Path): List[ClasspathEntry] = {
    Files
      .readAllLines(dependencies, StandardCharsets.UTF_8)
      .asScala
      .iterator
      .map(_.split("\t"))
      .collect { case Array(groupId, artifactId, version, entry) =>
        ClasspathEntry(
          entry = Paths.get(entry),
          sources = None,
          groupId = groupId,
          artifactId = artifactId,
          version = version
        )
      }
      .toList
  }

  /**
   * Parses ClasspathEntry from a "javacopts.txt" file in the targetroot.
   *
   * Every line of the file represents a Java compiler options, such as
   * "-classpath" or "-encoding".
   */
  private def fromJavacopts(
      javacopts: Path,
      sourceroot: Path
  ): List[ClasspathEntry] = {
    Files
      .readAllLines(javacopts, StandardCharsets.UTF_8)
      .asScala
      .iterator
      .map(_.stripPrefix("\"").stripSuffix("\""))
      .sliding(2)
      .collect {
        case Seq("-d", classesDirectory) =>
          fromClassesDirectory(Paths.get(classesDirectory), sourceroot).toList
        case Seq("-cp" | "-classpath", classpath) =>
          classpath
            .split(File.pathSeparator)
            .iterator
            .map(Paths.get(_))
            .flatMap(ClasspathEntry.fromClasspathJarFile)
            .toList
      }
      .flatten
      .toList
  }

  private def fromClassesDirectory(
      classesDirectory: Path,
      sourceroot: Path
  ): Option[ClasspathEntry] = {
    @tailrec
    def loop(dir: Path): Option[ClasspathEntry] = {
      if (dir == null || !dir.startsWith(sourceroot))
        None
      else
        fromPomXml(dir.resolve("pom.xml"), classesDirectory, None) match {
          case None =>
            loop(dir.getParent())
          case Some(value) =>
            Some(value)
        }
    }
    loop(classesDirectory.getParent())
  }

  /**
   * Tries to parse a ClasspathEntry from the POM file that lies next to the
   * given jar file.
   */
  private def fromClasspathJarFile(jar: Path): Option[ClasspathEntry] = {
    val pom = jar
      .resolveSibling(jar.getFileName.toString.stripSuffix(".jar") + ".pom")
    val sources = Option(
      jar.resolveSibling(
        jar.getFileName.toString.stripSuffix(".jar") + ".sources"
      )
    ).filter(Files.isRegularFile(_))
    fromPomXml(pom, jar, sources)
  }

  private def fromPomXml(
      pom: Path,
      classpathEntry: Path,
      sources: Option[Path]
  ): Option[ClasspathEntry] = {
    if (Files.isRegularFile(pom)) {
      val xml = scala.xml.XML.loadFile(pom.toFile)
      def xmlValue(key: String): String = {
        val node = xml \ key
        if (node.isEmpty)
          (xml \ "parent" \ key).text
        else
          node.text
      }.trim
      val groupId = xmlValue("groupId")
      val artifactId = xmlValue("artifactId")
      val version = xmlValue("version")
      Some(
        ClasspathEntry(classpathEntry, sources, groupId, artifactId, version)
      )
    } else {
      None
    }
  }
}

File: scip-java/src/main/kotlin/org/scip_code/scip_java/buildtools/ClasspathEntry.kt
package org.scip_code.scip_java.buildtools

import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import javax.xml.parsers.DocumentBuilderFactory
import org.scip_code.scip_java.aggregator.MavenPackage
import org.w3c.dom.Element
import org.w3c.dom.Node

/**
 * Represents a single classpath entry on the classpath of a project, used to emit SCIP
 * "packageInformation" nodes. A classpath entry can either be a jar file or a directory path.
 */
data class ClasspathEntry(
    val entry: Path,
    val groupId: String,
    val artifactId: String,
    val version: String,
) {
    fun mavenCoordinate(): String = "maven:$groupId:$artifactId:$version"

    fun toPackageInformation(): MavenPackage = MavenPackage(entry, groupId, artifactId, version)

    companion object {
        /**
         * Parses ClasspathEntry from the SCIP targetroot directory.
         *
         * Two separate formats are supported:
         * - javacopts.txt: line-separated list of Java compiler options.
         * - dependencies.txt: line-separated list of dependency information.
         *
         * Note that the targetroot can contain several files with names ending in
         * "dependencies.txt" - for example if they come from a multi-module build.
         */
        @JvmStatic
        fun fromTargetroot(targetroot: Path, sourceroot: Path): List<ClasspathEntry> {
            val javacopts = targetroot.resolve("javacopts.txt")
            return if (Files.isRegularFile(javacopts)) {
                fromJavacopts(javacopts, sourceroot)
            } else {
                discoverDependenciesFromFiles(targetroot)
            }
        }

        private fun discoverDependenciesFromFiles(targetroot: Path): List<ClasspathEntry> {
            if (!Files.isDirectory(targetroot)) return emptyList()
            return Files.list(targetroot).use { stream ->
                stream
                    .filter {
                        Files.isRegularFile(it) &&
                            it.fileName.toString().endsWith("dependencies.txt")
                    }
                    .toArray()
                    .map { it as Path }
                    .flatMap { fromDependencies(it) }
                    .distinct()
            }
        }

        private fun fromDependencies(dependencies: Path): List<ClasspathEntry> =
            Files.readAllLines(dependencies, StandardCharsets.UTF_8).mapNotNull { line ->
                val parts = line.split('\t')
                if (parts.size == 4) {
                    val (groupId, artifactId, version, entry) = parts
                    ClasspathEntry(
                        entry = Paths.get(entry),
                        groupId = groupId,
                        artifactId = artifactId,
                        version = version,
                    )
                } else null
            }

        private fun fromJavacopts(javacopts: Path, sourceroot: Path): List<ClasspathEntry> {
            val lines =
                Files.readAllLines(javacopts, StandardCharsets.UTF_8).map {
                    it.removePrefix("\"").removeSuffix("\"")
                }
            val result = mutableListOf<ClasspathEntry>()
            for (i in 0 until lines.size - 1) {
                val key = lines[i]
                val value = lines[i + 1]
                when (key) {
                    "-d" -> {
                        val entry = fromClassesDirectory(Paths.get(value), sourceroot)
                        if (entry != null) result += entry
                    }
                    "-cp",
                    "-classpath" -> {
                        value.split(File.pathSeparator).forEach { jarPath ->
                            val entry = fromClasspathJarFile(Paths.get(jarPath))
                            if (entry != null) result += entry
                        }
                    }
                }
            }
            return result
        }

        private tailrec fun fromClassesDirectory(
            classesDirectory: Path,
            sourceroot: Path,
            currentDir: Path? = classesDirectory.parent,
        ): ClasspathEntry? {
            if (currentDir == null || !currentDir.startsWith(sourceroot)) return null
            val pomEntry = fromPomXml(currentDir.resolve("pom.xml"), classesDirectory)
            if (pomEntry != null) return pomEntry
            return fromClassesDirectory(classesDirectory, sourceroot, currentDir.parent)
        }

        /**
         * Tries to parse a ClasspathEntry from the POM file that lies next to the given jar file.
         */
        private fun fromClasspathJarFile(jar: Path): ClasspathEntry? {
            val fileName = jar.fileName?.toString() ?: return null
            val base = fileName.removeSuffix(".jar")
            val pom = jar.resolveSibling("$base.pom")
            return fromPomXml(pom, jar)
        }

        private fun fromPomXml(pom: Path, classpathEntry: Path): ClasspathEntry? {
            if (!Files.isRegularFile(pom)) return null
            val factory =
                DocumentBuilderFactory.newInstance().apply {
                    isNamespaceAware = false
                    isValidating = false
                    // Defensive: disable external entity resolution to avoid
                    // accidental XXE against attacker-controlled pom.xml files.
                    setFeature("http://apache.org/xml/features/disallow-doctype-decl", true)
                    setFeature("http://xml.org/sax/features/external-general-entities", false)
                    setFeature("http://xml.org/sax/features/external-parameter-entities", false)
                    isXIncludeAware = false
                    isExpandEntityReferences = false
                }
            val document = factory.newDocumentBuilder().parse(Files.newInputStream(pom))
            val root = document.documentElement ?: return null

            fun textOf(parent: Element, key: String): String {
                val direct = childElementByName(parent, key)
                if (direct != null) return direct.textContent.orEmpty().trim()
                val parentTag = childElementByName(parent, "parent") ?: return ""
                return childElementByName(parentTag, key)?.textContent.orEmpty().trim()
            }

            return ClasspathEntry(
                entry = classpathEntry,
                groupId = textOf(root, "groupId"),
                artifactId = textOf(root, "artifactId"),
                version = textOf(root, "version"),
            )
        }

        private fun childElementByName(parent: Element, name: String): Element? {
            val nodes = parent.childNodes
            for (i in 0 until nodes.length) {
                val node = nodes.item(i)
                if (node.nodeType == Node.ELEMENT_NODE && (node as Element).tagName == name) {
                    return node
                }
            }
            return null
        }
    }
}
